File: backend/app/models/complaint.py
import uuid
from sqlalchemy import Column, String, Text, DateTime, ForeignKey, Enum
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.core.database import Base
import enum


class ComplaintCategory(str, enum.Enum):
    noise = "noise"
    safety = "safety"
    billing = "billing"
    neighbor = "neighbor"
    management = "management"
    other = "other"


class ComplaintStatus(str, enum.Enum):
    open = "open"
    under_review = "under_review"
    resolved = "resolved"
    dismissed = "dismissed"


class Complaint(Base):
    __tablename__ = "complaints"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    unit_id = Column(UUID(as_uuid=True), ForeignKey("units.id", ondelete="CASCADE"), nullable=False, index=True)
    tenant_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True)
    category = Column(Enum(ComplaintCategory), nullable=False)
    subject = Column(String(200), nullable=False)
    description = Column(Text, nullable=False)
    status = Column(Enum(ComplaintStatus), nullable=False, default=ComplaintStatus.open, index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    unit = relationship("Unit", back_populates="complaints")
    tenant = relationship("User", foreign_keys=[tenant_id])
    responses = relationship("ComplaintResponse", back_populates="complaint", cascade="all, delete-orphan")


class ComplaintResponse(Base):
    __tablename__ = "complaint_responses"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    complaint_id = Column(UUID(as_uuid=True), ForeignKey("complaints.id", ondelete="CASCADE"), nullable=False, index=True)
    responder_id = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=False)
    message = Column(Text, nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    complaint = relationship("Complaint", back_populates="responses")
    responder = relationship("User", foreign_keys=[responder_id])

File: frontend/src/types/complaint.ts
export type ComplaintCategory =
  | 'noise'
  | 'safety'
  | 'billing'
  | 'neighbor'
  | 'management'
  | 'other';

export type ComplaintStatus =
  | 'open'
  | 'under_review'
  | 'resolved'
  | 'dismissed';

export interface UserInfo {
  id: string;
  email: string;
  full_name: string | null;
}

export interface ComplaintResponse {
  id: string;
  complaint_id: string;
  responder_id: string;
  message: string;
  created_at: string;
  responder: UserInfo | null;
}

export interface Complaint {
  id: string;
  unit_id: string;
  tenant_id: string;
  category: ComplaintCategory;
  subject: string;
  description: string;
  status: ComplaintStatus;
  created_at: string;
  updated_at: string;
  tenant: UserInfo | null;
  responses: ComplaintResponse[];
}

export interface ComplaintCreate {
  unit_id: string;
  category: ComplaintCategory;
  subject: string;
  description: string;
}

export interface ComplaintUpdate {
  status?: ComplaintStatus;
}

export interface ComplaintResponseCreate {
  message: string;
}